### ISSUE TITLE
Support for temporary STS credentials?

### ISSUE THREAD

**Issue by garnaat** [tedder/requests-aws4auth#11]
More of a question than an issue, really.  Does this currently support STS temporary credentials?  If not, any plans for that?

**Comment by magdalene** [comment]
@garnaat -- this is a feature I needed, so I added it: https://github.com/sam-washington/requests-aws4auth/pull/12

I'm very new using AWS, so I have no idea honestly whether it works for other people's use cases, but it is working for me (I'm using it in a Lambda with an IAM role-based policy to call an AWS api).

**Comment by garnaat** [comment]
I'll give it a try. Thanks!

**Comment by magdalene** [comment]
If you want to try it, in case it's not clear from the changes, I'm calling the constructor like this:

```
AWS4Auth(
    '<ACCESS_KEY>',
    '<SECRET_ACCESS>',
    'eu-west-1',
    's3',
    include_hdrs=AWS4Auth.default_include_headers + ['x-amz-security-token'],
    session_token='<SESSION_TOKEN>')
```

**Comment by lingrlongr** [comment]
I'm trying your implementation @magdalene with something like:

```
from boto.provider import get_default
import elasticsearch
from requests_aws4auth import AWS4Auth

p = get_default()
host = '<MY_HOST>'

awsauth = AWS4Auth(
    p.get_access_key(),
    p.get_secret_key(),
    'us-east-1',
    'es',
    include_hdrs=AWS4Auth.default_include_headers + ['x-amz-security-token'],
    session_token=p.get_security_token())

s = elasticsearch.Elasticsearch(
    hosts=[{'host': host, 'port': 443}],
    http_auth=awsauth,
    use_ssl=True,
    verify_certs=True,
    connection_class=elasticsearch.connection.RequestsHttpConnection
)

print es.cluster.health()
```

But when I run it, I get:

```
raise HTTP_EXCEPTIONS.get(status_code, TransportError)(status_code, error_message, additional_info)
elasticsearch.exceptions.AuthorizationException: TransportError(403, u'{"message":"The security token included in the request is invalid."}')
```

It's retrieving credentials correctly from instance metadata (from a role), but they don't seem to work.  Any idea why that would be?

**Comment by magdalene** [comment]
@lingrlongr The only time I have seen that specific message is when I was legitimately [accidentally] specifying a wrong session token. I haven't used `boto.provider.get_default` to retrieve the credentials, instead I have used `botocore.credentials.get_credentials` (botocore is a dependency of  boto3), and I've also tested explicitly retrieving credentials from the sts service. Assuming you've confirmed that the session_token looks reasonable (it's not accidentally None), I'm not sure what the problem is. I am pasting the code I'm using though in case it helps you:

(Not specifying the extra header anymore in the example below because I'm using this https://github.com/sam-washington/requests-aws4auth/pull/13 -- but I used it with the old version as well)

```
from botocore.credentials import get_credentials
from botocore.session import Session
from elasticsearch import Elasticsearch, RequestsHttpConnection
creds = get_credentials(Session())
awsauth = AWS4Auth(
        creds.access_key,
        creds.secret_key,
    'eu-west-1',
    'es',
    session_token=creds.token)
es = Elasticsearch(
        hosts=[{'host':'<ES_HOST>', 'port': 443}],
        http_auth=awsauth,
        use_ssl=True,
        verify_certs=True,
        connection_class=RequestsHttpConnection)
# proceed to use the es client
```

The places I am using what is above are in an AWS Lambda function (where the keys/tokens end up coming from environment variables), and on an EC2 instance which has that role (which ends up calling the metadata service to get the credentials). 

On my own dev machine I've also tested by retrieving temporary credentials from the sts service explicitly, something like:

```
import boto3
client = boto3.client('sts')
creds = client.assume_role(
        RoleArn='<ROLE_ARN>',
        RoleSessionName='<A STRING I MADE UP>'))['Credentials']
awsauth = AWS4Auth(
        creds['AccessKeyId'],
        creds['SecretAccessKey'],
    'eu-west-1',
    'es',
    session_token=creds[['SessionToken'])
# construct es client as above
```

Last comment -- I would think the issue is actually with the session token, NOT with one of the other values, because I was getting a different error message when I did various other wrong things.

**Comment by lingrlongr** [comment]
Thanks for the reply @magdalene.  I'm pretty sure I found the problem.  I think the version that's on pypi may be out of date.  If I clone the git repo and install from there `pip install -e .` into my virtualenv, it works.  Then it all works!

**Comment by lingrlongr** [comment]
@sam-washington - When will latest code here be released to pypi?

**Comment by sam-washington** [comment]
Sorry @lingrlongr, I've been really strapped for time recently. Things should be improving soon though! I'll update PyPi this weekend.

**Comment by lingrlongr** [comment]
Thanks @sam-washington :)

**Comment by sam-washington** [comment]
Hi, the changes have been merged to master, and released on github and pypi as version 0.9

**Comment by lingrlongr** [comment]
Awesome.  Thanks @sam-washington !
